File: apps/go/design-pattern/templatemethod/templatemethod_test.go
package templatemethod

import (
	"strings"
	"testing"
)

func TestHtmlReport(t *testing.T) {
	output := GenerateReport(HtmlFormat(), "月次報告", []string{"順調", "最高の調子"})

	if !strings.Contains(output, "<html>") {
		t.Error("HTML 出力に <html> が含まれていない")
	}
	if !strings.Contains(output, "<title>月次報告</title>") {
		t.Error("HTML 出力にタイトルが含まれていない")
	}
	if !strings.Contains(output, "<p>順調</p>") {
		t.Error("HTML 出力に本文が含まれていない")
	}
	if !strings.Contains(output, "</html>") {
		t.Error("HTML 出力に </html> が含まれていない")
	}
}

func TestPlainTextReport(t *testing.T) {
	output := GenerateReport(PlainTextFormat(), "月次報告", []string{"順調", "最高の調子"})

	if !strings.Contains(output, "**** 月次報告 ****") {
		t.Error("プレーンテキスト出力にタイトルが含まれていない")
	}
	if !strings.Contains(output, "順調") {
		t.Error("プレーンテキスト出力に本文が含まれていない")
	}
	if strings.Contains(output, "<html>") {
		t.Error("プレーンテキスト出力に HTML タグが含まれている")
	}
}

func TestPlainTextReportContainsAllLines(t *testing.T) {
	output := GenerateReport(PlainTextFormat(), "月次報告", []string{"順調", "最高の調子"})

	if !strings.Contains(output, "最高の調子") {
		t.Error("プレーンテキスト出力にすべての行が含まれていない")
	}
}

func TestHtmlReportStructure(t *testing.T) {
	output := GenerateReport(HtmlFormat(), "テスト", []string{"行1"})
	lines := strings.Split(output, "\n")

	if lines[0] != "<html>" {
		t.Errorf("最初の行が <html> ではない: %s", lines[0])
	}
	if lines[len(lines)-1] != "</html>" {
		t.Errorf("最後の行が </html> ではない: %s", lines[len(lines)-1])
	}
}

func TestEmptyText(t *testing.T) {
	output := GenerateReport(HtmlFormat(), "空レポート", []string{})

	if !strings.Contains(output, "<html>") {
		t.Error("空のテキストでも HTML 構造が出力されるべき")
	}
	if !strings.Contains(output, "<title>空レポート</title>") {
		t.Error("空のテキストでもタイトルは出力されるべき")
	}
}

func TestCustomFormat(t *testing.T) {
	markdown := ReportFormat{
		OutputHead: func(title string) []string {
			return []string{"# " + title, ""}
		},
		OutputLine: func(line string) []string {
			return []string{"- " + line}
		},
	}

	output := GenerateReport(markdown, "報告", []string{"項目1", "項目2"})

	if !strings.Contains(output, "# 報告") {
		t.Error("カスタムフォーマットのタイトルが正しくない")
	}
	if !strings.Contains(output, "- 項目1") {
		t.Error("カスタムフォーマットの行が正しくない")
	}
}

File: apps/haskell/design-pattern/04-template-method/test/TemplateMethodTest.hs
module TemplateMethodTest (tests) where

import Data.List (isInfixOf)
import TemplateMethod (ReportFormat (..), generateReport, htmlFormat, plainTextFormat)
import Test.HUnit

tests :: Test
tests =
  TestLabel "TemplateMethodTest" $
    TestList
      [ TestLabel "testHtmlFormat" testHtmlFormat
      , TestLabel "testPlainTextFormat" testPlainTextFormat
      , TestLabel "testCustomFormat" testCustomFormat
      ]

testHtmlFormat :: Test
testHtmlFormat = TestCase $ do
  let result = generateReport htmlFormat "月次報告" ["順調", "問題なし"]
  assertBool "HTML 開始タグを含む" ("<html>" `isInfixOf` result)
  assertBool "タイトルを含む" ("<title>月次報告</title>" `isInfixOf` result)
  assertBool "本文行を含む" ("<p>順調</p>" `isInfixOf` result)
  assertBool "HTML 終了タグを含む" ("</html>" `isInfixOf` result)

testPlainTextFormat :: Test
testPlainTextFormat = TestCase $ do
  let result = generateReport plainTextFormat "月次報告" ["順調", "問題なし"]
  assertEqual
    "プレーンテキスト形式"
    "**** 月次報告 ****\n\n順調\n問題なし\n"
    result

testCustomFormat :: Test
testCustomFormat = TestCase $ do
  let csvFormat =
        ReportFormat
          { outputStart = \title -> title ++ "\n"
          , outputLine = \line -> line ++ ","
          , outputEnd = \_ -> "\n"
          }
  let result = generateReport csvFormat "レポート" ["A", "B", "C"]
  assertEqual "カスタムフォーマットを追加できる" "レポート\nA,B,C,\n" result
